File: web/src/types/produce.ts
export type ListingStatus = 'available' | 'sold' | 'reserved';
export type TransactionStatus = 'pending' | 'quoted' | 'confirmed' | 'completed' | 'cancelled';
export type BuyerQuoteAction = 'approve' | 'switch_pickup' | 'reject';

export interface ProduceCategoryConfig {
  id: string;
  name: string;
  label: string;
  icon: string;
  suggestedUnits: string[];
  samplePlaceholder: string;
}

export const PRODUCE_CATEGORIES: ProduceCategoryConfig[] = [
  {
    id: 'vegetables',
    name: 'Vegetables',
    label: 'Vegetables (Gulay)',
    icon: '🥬',
    suggestedUnits: ['kg', 'sack (50kg)', 'crate / kaing', 'bundle / tali', 'box'],
    samplePlaceholder: 'e.g. Red Tomato (Kamatis), Eggplant (Talong), Cabbage',
  },
  {
    id: 'fruits',
    name: 'Fruits',
    label: 'Fruits (Prutas)',
    icon: '🍌',
    suggestedUnits: ['kg', 'crate / kaing', 'box', 'piece', 'ton (MT)'],
    samplePlaceholder: 'e.g. Carabao Mango (Mangga), Lakatan Banana, Papaya',
  },
  {
    id: 'grains',
    name: 'Grains & Cereals',
    label: 'Grains & Cereals (Palay / Mais)',
    icon: '🌾',
    suggestedUnits: ['kg', 'sack (50kg)', 'cavan', 'ton (MT)'],
    samplePlaceholder: 'e.g. Yellow Corn (Mais), Dry Palay (Paddy Rice), Milled Rice',
  },
  {
    id: 'root_crops',
    name: 'Root Crops',
    label: 'Root Crops & Tubers (Kamote / Cassava)',
    icon: '🥔',
    suggestedUnits: ['kg', 'sack (50kg)', 'crate', 'ton (MT)'],
    samplePlaceholder: 'e.g. Fresh Cassava, Sweet Potato (Camote), Ube, Taro (Gabi)',
  },
  {
    id: 'livestock',
    name: 'Livestock & Poultry',
    label: 'Livestock & Poultry (Baka, Baboy, Manok, Kambing)',
    icon: '🐓',
    suggestedUnits: ['head', 'live weight kg', 'dressed kg', 'pair', 'tray (30 eggs)', 'piece'],
    samplePlaceholder: 'e.g. Free-range Native Chicken, Boer Goat (Kambing), Fattened Swine (Baboy), Cattle',
  },
  {
    id: 'fisheries',
    name: 'Fisheries & Aquaculture',
    label: 'Fisheries & Aquaculture (Isda / Hipon)',
    icon: '🐟',
    suggestedUnits: ['kg', 'styro box (banyera)', 'bucket', 'piece'],
    samplePlaceholder: 'e.g. Fresh Tilapia, Bangus (Milkfish), Catfish (Hito), Tiger Prawn',
  },
  {
    id: 'spices',
    name: 'Spices & Herbs',
    label: 'Spices & Herbs (Pampalasa)',
    icon: '🌶️',
    suggestedUnits: ['kg', 'bundle / tali', 'sack', 'pack'],
    samplePlaceholder: 'e.g. Red Onion (Sibuyas), Native Garlic (Bawang), Ginger (Luya), Siling Labuyo',
  },
  {
    id: 'processed',
    name: 'Agri-Processed & By-products',
    label: 'Agri-Processed & By-products (Kape, Asukal, Honey)',
    icon: '🍯',
    suggestedUnits: ['kg', 'bottle', 'jar', 'pack', 'liter'],
    samplePlaceholder: 'e.g. Pure Honey, Tablea Cacao, Robusta Coffee Beans, Muscovado Sugar',
  },
];

export interface ProduceListing {
  id: string;
  farmerId: string;
  farmerName: string;
  farmerPhone?: string;
  cropName: string;
  category: string;
  quantity: number;
  unit: string;
  pricePerUnit: number;
  harvestDate: string;
  location: string;
  photos?: string[];
  description?: string;
  status: ListingStatus;
  createdAt: string;
  updatedAt: string;
}

export interface CreateProduceListingPayload {
  cropName: string;
  category: string;
  quantity: number;
  unit: string;
  pricePerUnit: number;
  harvestDate: string;
  location: string;
  photos?: string[];
  description?: string;
  status?: ListingStatus;
}

export interface ProduceTransaction {
  id: string;
  listingId: string;
  cropName: string;
  cropPhoto?: string;
  buyerId: string;
  buyerName: string;
  farmerId: string;
  farmerName: string;
  quantity: number;
  unitPrice: number;
  subtotal?: number;
  shippingFee?: number;
  totalPrice: number;
  deliveryMethod?: string;
  deliveryAddress?: string;
  contactMessage?: string;
  status: TransactionStatus;
  createdAt: string;
  updatedAt: string;
}

export interface CreateProduceTransactionPayload {
  listingId: string;
  quantity: number;
  contactMessage?: string;
  deliveryMethod?: string;
  deliveryAddress?: string;
}

File: backend/internal/models/produce.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/v2/bson"
)

type ListingStatus string

const (
	ListingAvailable ListingStatus = "available"
	ListingSold      ListingStatus = "sold"
	ListingReserved  ListingStatus = "reserved"
)

type TransactionStatus string

const (
	TxPending   TransactionStatus = "pending"
	TxConfirmed TransactionStatus = "confirmed"
	TxCompleted TransactionStatus = "completed"
	TxCancelled TransactionStatus = "cancelled"
)

// ProduceListing represents a farmer's crop listing.
type ProduceListing struct {
	ID           bson.ObjectID `bson:"_id,omitempty"       json:"id"`
	FarmerID     bson.ObjectID `bson:"farmer_id"           json:"farmerId"`
	FarmerName   string        `bson:"farmer_name"         json:"farmerName"`
	FarmerPhone  string        `bson:"farmer_phone"        json:"farmerPhone,omitempty"`
	CropName     string        `bson:"crop_name"           json:"cropName"`
	Category     string        `bson:"category"            json:"category"`
	Quantity     float64       `bson:"quantity"            json:"quantity"`
	Unit         string        `bson:"unit"                json:"unit"` // e.g. "kg", "tons", "sacks"
	PricePerUnit float64       `bson:"price_per_unit"      json:"pricePerUnit"`
	HarvestDate  time.Time     `bson:"harvest_date"        json:"harvestDate"`
	Location     string        `bson:"location"            json:"location"` // Region / Location
	Photos       []string      `bson:"photos,omitempty"    json:"photos,omitempty"`
	Description  string        `bson:"description,omitempty" json:"description,omitempty"`
	Status       ListingStatus `bson:"status"              json:"status"`
	CreatedAt    time.Time     `bson:"created_at"          json:"createdAt"`
	UpdatedAt    time.Time     `bson:"updated_at"          json:"updatedAt"`
}

// CreateProduceListingRequest is the JSON payload to create a produce listing.
type CreateProduceListingRequest struct {
	CropName     string    `json:"cropName"`
	Category     string    `json:"category"`
	Quantity     float64   `json:"quantity"`
	Unit         string    `json:"unit"`
	PricePerUnit float64   `json:"pricePerUnit"`
	HarvestDate  string    `json:"harvestDate"` // YYYY-MM-DD
	Location     string    `json:"location"`
	Photos       []string  `json:"photos,omitempty"`
	Description  string    `json:"description,omitempty"`
}

// UpdateProduceListingRequest is the JSON payload to update a produce listing.
type UpdateProduceListingRequest struct {
	CropName     *string        `json:"cropName,omitempty"`
	Category     *string        `json:"category,omitempty"`
	Quantity     *float64       `json:"quantity,omitempty"`
	Unit         *string        `json:"unit,omitempty"`
	PricePerUnit *float64       `json:"pricePerUnit,omitempty"`
	HarvestDate  *string        `json:"harvestDate,omitempty"`
	Location     *string        `json:"location,omitempty"`
	Photos       []string       `json:"photos,omitempty"`
	Description  *string        `json:"description,omitempty"`
	Status       *ListingStatus `json:"status,omitempty"`
}

// ProduceTransaction represents a purchase order for produce.
type ProduceTransaction struct {
	ID             bson.ObjectID     `bson:"_id,omitempty"       json:"id"`
	ListingID      bson.ObjectID     `bson:"listing_id"          json:"listingId"`
	CropName       string            `bson:"crop_name"           json:"cropName"`
	CropPhoto      string            `bson:"crop_photo,omitempty" json:"cropPhoto,omitempty"`
	BuyerID        bson.ObjectID     `bson:"buyer_id"            json:"buyerId"`
	BuyerName      string            `bson:"buyer_name"          json:"buyerName"`
	FarmerID       bson.ObjectID     `bson:"farmer_id"           json:"farmerId"`
	FarmerName     string            `bson:"farmer_name"         json:"farmerName"`
	Quantity       float64           `bson:"quantity"            json:"quantity"`
	UnitPrice      float64           `bson:"unit_price"          json:"unitPrice"`
	Subtotal       float64           `bson:"subtotal"            json:"subtotal"`
	ShippingFee    float64           `bson:"shipping_fee"        json:"shippingFee"`
	TotalPrice     float64           `bson:"total_price"         json:"totalPrice"`
	DeliveryMethod string            `bson:"delivery_method,omitempty" json:"deliveryMethod,omitempty"`
	DeliveryAddress string           `bson:"delivery_address,omitempty" json:"deliveryAddress,omitempty"`
	ContactMessage string            `bson:"contact_message,omitempty" json:"contactMessage,omitempty"`
	Status         TransactionStatus `bson:"status"              json:"status"`
	CreatedAt      time.Time         `bson:"created_at"          json:"createdAt"`
	UpdatedAt      time.Time         `bson:"updated_at"          json:"updatedAt"`
}

// CreateProduceTransactionRequest is the JSON body to initiate a purchase request.
type CreateProduceTransactionRequest struct {
	ListingID       string  `json:"listingId"`
	Quantity        float64 `json:"quantity"`
	ContactMessage  string  `json:"contactMessage,omitempty"`
	DeliveryMethod  string  `json:"deliveryMethod,omitempty"`
	DeliveryAddress string  `json:"deliveryAddress,omitempty"`
}

// UpdateTransactionStatusRequest is the JSON payload to update transaction status.
type UpdateTransactionStatusRequest struct {
	Status      TransactionStatus `json:"status"`
	ShippingFee *float64          `json:"shippingFee,omitempty"`
}
